src/results/bulk_write.rs

```diff
@@ -0,0 +1,85 @@
+#![allow(missing_docs)]
+
+use std::{collections::HashMap, fmt::Debug};
+
+use serde::Serialize;
+use serde_with::skip_serializing_none;
+
+use crate::{
+    results::{DeleteResult, InsertOneResult, UpdateResult},
+    serde_util::serialize_indexed_map,
+};
+
+#[skip_serializing_none]
+#[derive(Clone, Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+#[non_exhaustive]
+pub struct BulkWriteResult {
```

Comment:
My original design for bulk write results included both a `SummaryBulkWriteResult` and `VerboseBulkWriteResult`, with only the latter containing the result maps. In that design, the `bulk_write` method returned a `BulkWrite<SummaryBulkWriteResult>`, and that type had a `verbose_results` method that took no parameters and returned a `BulkWrite<VerboseBulkWriteResult>`. The action implementations then returned the result type present on the action struct. I still think this alternative design is preferable as it's a little awkward to return a result struct that will always contain empty maps in the default case, but it got a little finicky to implement the `partial_result` field of `BulkWriteError`. I think it would make more sense to do this as follow-on work rather than in this PR to keep things simpler here; I can file a ticket if that sounds good.

Reply:
Yup, that sounds good.